src/core/quantum_gate.jl

```diff
@@ -101,6 +133,106 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
+julia> controlled_hadamard=ControlledGate(hadamard,[1,2])
+Gate Object: ControlledGate{Snowflake.Hadamard}
+Connected_qubits	: [1, 2]
+Operator:
+(4, 4)-element Snowflake.DenseOperator:
+Underlying data ComplexF64:
+1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    0.7071067811865475 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    -0.7071067811865475 + 0.0im
+
+```
+
+It can then be used in a `QuantumCircuit` as any other `Gate`, and its display symbol is 
+inherited from the display symbol of the single-target `Hadamard` `Gate`:
+
+```jldoctest controlled_hadamard
+julia> circuit=QuantumCircuit(qubit_count=2,gates=[controlled_hadamard])
+Quantum Circuit Object:
+   qubit_count: 2 
+q[1]:──*──
+       |  
+q[2]:──H──
+          
+```
+
+
+"""
+struct ControlledGate{G<:AbstractGate}<:AbstractControlledGate
+    kernel::AbstractOperator
+    connected_qubits::Vector{Int}
+
+    function ControlledGate(
+        kernel_handle::Function,
+        connected_qubits::Vector{Int};
+        params::Vector{<:Real}=Vector{Float64}(),
+        precision::Integer=4)
+
+        @assert length(connected_qubits)==2 "Not implemented for multi-control kernel Operators"
+
+        @assert connected_qubits[1]!=connected_qubits[2] "Control and target qubit must be different, received $connected_qubits"
+
+        @assert haskey(operator_to_gate_mapping,kernel_handle) "Function $kernel_handle is not associated with a Gate constructor"
+
+        # calls function, returns Operator
+        kernel=kernel_handle(params...) 
+
+        # Gate type associated with this operator is used to distinguish between 
+        # ControlledGate types and to assign a display symbol to them
+        kernel_gate_type=operator_to_gate_mapping[kernel_handle]  
+
+        # build new display symbol using existing symbol pertaining to kernel
+
+        symbol_specs=gates_display_symbols[kernel_gate_type]      
+        num_params=length(params)
+        gate_params=Dict{String,Real}() 
+        
+        for (field,param) in zip(symbol_specs[end-num_params+1:end],params)
+            gate_params[field]=param
+        end
+
+        gates_display_symbols[ControlledGate{kernel_gate_type}]=[
+            control_display_symbol,
+            format_label(
+                symbol_specs,
+                1, 
+                gate_params;
+                precision=precision
+            )[1]
+        ]
+        
+        new{kernel_gate_type}(kernel,connected_qubits)
+    end
+end
+
+# returns an equivalent size 4 DenseOperator built using the kernel
+function get_operator(gate::ControlledGate, T::Type{<:Complex}=ComplexF64) 
+    op=get_matrix(eye(4))
+    
+    op[3:4,3:4]=get_matrix(gate.kernel)
+
+    return DenseOperator(convert(Matrix{T},op))
+end
+
+get_connected_qubits(gate::ControlledGate)=gate.connected_qubits
```

Comment:
The gate will only contain the target qubits, but connected qubits should also include the control qubits.

Reply:
@plerouxanyon  the `connected_qubits` passed at construction of ControlledGate has both